### DISCUSSION TITLE
第19题 接受ETH 死活通不过

### DISCUSSION THREAD

**Discussion by flybottle** [AmazingAng/WTF-Solidity#455] [General]
vitalik想部署一个只接收ETH同时不需要接收msg.data的合约，那么他部署的合约中必须有什么函数？

这道题目的题意我就没弄明白，vitalik部署的这个合约是不允许别人传送msg.data还是说别人可以传msg.data,但合约也能处理？

**Comment by maixuanfeng1** [comment]
这题现在好像已经改了

**Comment by nshen** [comment]
我也过不去了，穷举了半天答案应该是B

<img width="1101" alt="image" src="https://github.com/AmazingAng/WTF-Solidity/assets/181506/66d81a47-73f7-4e12-bd16-49a485c53fb0">

为什么不是D呢？ 任意一个不都能接收吗？

```
    Ether is sent to contract
        is msg.data empty?
            /   \
            yes  no
            /     \
    receive()?  fallback()
        /   \
     yes    no
    /        \
receive()  fallback()
```

**Reply by XdpCs** [reply]
The fallback function is executed on a call to the contract if none of the other functions match the given function signature, or if no data was supplied at all and there is no [receive Ether function](https://docs.soliditylang.org/en/v0.8.22/contracts.html#receive-ether-function). The fallback function always receives data, but in order to also receive Ether it must be marked payable.
在Solidity官方文档中，说明了上述的关系，教程的描述没有问题。

**Reply by XdpCs** [reply]
refer: https://docs.soliditylang.org/en/v0.8.22/contracts.html#receive-ether-function

**Reply by XdpCs** [reply]
因为他需要传msg.data。

**Reply by nshen** [reply]
receive 不能接收msg.data 吗？

**Reply by XdpCs** [reply]
是的

**Reply by XdpCs** [reply]
你如果尝试在receive中 使用msg.data 会报下面的错误
CompileError: TypeError: "msg.data" cannot be used inside of "receive" function.
